### Heroes Of Pymoli Data Analysis
* Of the 1163 active players, the vast majority are male (84%). There also exists, a smaller, but notable proportion of female players (14%).

* Our peak age demographic falls between 20-24 (44.8%) with secondary groups falling between 15-19 (18.60%) and 25-29 (13.4%).  
-----

### Note
* Instructions have been included for each segment. You do not have to follow them exactly, but they are included to help you think through the steps.

In [298]:
# Dependencies and Setup
import pandas as pd
import numpy as np

# File to Load (Remember to Change These)
file_to_load = "Resources/purchase_data.csv"

# Read Purchasing File and store into Pandas data frame
purchase_data = pd.read_csv(file_to_load)

# look for missing values and general data in data for possible cleanup
#purchase_data.count()
#purchase_data.columns
#purchase_data.describe()
#purchase_data.index
#purchase_data.style
#purchase_data.dtypes
#purchase_data.head()
###check_for_NA
#purchase_data.isna().sum()   # or s.isnull().sum() for older pandas versions
# Remove the rows with missing data
# clean_ufo_df = ufo_df.dropna(how="any")
# clean_ufo_df.count()


## Player Count  - complete

* Display the total number of players


In [299]:
#Total = purchase_data.count()
#purchase_data.dtypes
#purchase_data.groupby('SN').sum()
clean_purchases = purchase_data.drop_duplicates(["SN", "Gender"])[["SN", "Gender"]]
Total = clean_purchases["Gender"].count()
print('The total Number of Players is', Total)
#clean_purchases.head()

The total Number of Players is 576


## Purchasing Analysis (Total) - complete

* Run basic calculations to obtain number of unique items, average price, etc.

* Create a summary data frame to hold the results


* Optional: give the displayed data cleaner formatting


* Display the summary data frame


In [300]:
Unique_item_count = purchase_data['Item ID'].nunique()
avg_price = purchase_data['Price'].mean()
Total_Revenue = purchase_data['Price'].sum()
purc_count = purchase_data['Purchase ID'].count()

# Improve formatting before outputting spreadsheet
summary_table_purch = pd.DataFrame({"Total Unique items": Unique_item_count,
                            "Average price": [avg_price],
                            "Number of Purchases":[purc_count],
                            "Total Revenue":[Total_Revenue]
                              })
# Improve formatting before outputting spreadsheet
summary_table_purch["Average price"] = summary_table_purch["Average price"].map( 
    "${:.2f}".format)
summary_table_purch["Total Revenue"] = summary_table_purch["Total Revenue"].map( 
    "${:.2f}".format)
summary_table_purch

Total Unique items Average price  Number of Purchases Total Revenue
0                 183         $3.05                  780      $2379.77

## Gender Demographics  - complete

* Percentage and Count of Male Players


* Percentage and Count of Female Players


* Percentage and Count of Other / Non-Disclosed
	Total Count	Percentage of Players
Male	484	84.03
Female	81	14.06
Other / Non-Disclosed	11	1.91



In [301]:
# print(count)
#Need unique dataset (try drop_duplicates) for SN and gender
#moved to top  #####clean_purchases = purchase_data.drop_duplicates(["SN", "Gender"])[["SN", "Gender"]]
#keeper###male_df = purchase_data.loc[(
#keeper###    purchase_data["Gender"] == "Male")]
#keeper###male_total = male_df["Gender"].count()
#keeper###print(male_total)
#moved to top  #####clean_purchases = clean_purchases.count()
count = clean_purchases["Gender"].value_counts()
percentage = count/Total
summary_table_gender = pd.DataFrame({"Total User by Gender": count,
                              "Percentage of Players": percentage,
                              })
# Improve formatting before outputting
summary_table_gender["Percentage of Players"] = summary_table_gender["Percentage of Players"].map( 
    "{:.2%}".format)

summary_table_gender

Total User by Gender Percentage of Players
Male                                    484                84.03%
Female                                   81                14.06%
Other / Non-Disclosed                    11                 1.91%


## Purchasing Analysis (Gender) - Not complete

* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. by gender
* Create a summary data frame to hold the results
* Optional: give the displayed data cleaner formatting
* Display the summary data frame
	Purchase Count	Average Purchase Price	Total Purchase Value	Avg Total Purchase per Person
    
Gender				
Female	113	$3.20	$361.94	$4.47
Male	652	$3.02	$1,967.64	$4.07
Other / Non-Disclosed	15	$3.35	$50.19	$4.56

In [302]:
# Using GroupBy in order to separate the data into fields according to "gender" values
grouped_gender_df = purchase_data.groupby(['Gender'])
grouped_gender_df.head()
#Purch_counts = grouped_gender_df["Purchase ID"].counts()
#print(Purchase_counts)

male_total = grouped_gender_df["Price"].sum()
#print(male_total)
male_avg_price = grouped_gender_df["Price"].mean()

#Avg Total Purchase per Person
avg_ttl = male_total / Total

#Creating a new DataFrame using both duration and count
male_summary_table = pd.DataFrame({"Total Purchase Value": male_total,
# #      error!?!?!?                                    "Purchase Count": Purch_counts,
                                    "Avg Total Purchase per Person": avg_ttl,
                                    "Average Purchase Price": male_avg_price})
male_summary_table["Average Purchase Price"] = male_summary_table["Average Purchase Price"].map( 
    "${:.2}".format)
male_summary_table["Avg Total Purchase per Person"] = male_summary_table["Avg Total Purchase per Person"].map( 
    "{:.2}".format)
male_summary_table["Total Purchase Value"] = male_summary_table["Total Purchase Value"].map( 
    "{:.2}".format)
male_summary_table.head()
#discarded efforts - -  below ###############
# male_df = purchase_data.loc[(
#       purchase_data["Gender"] == "Male")]
# #male_df = purchase_data.groupby('Gender')
# # for 'Gender', male_df in purchase_data:
# # male_df
# male_total = male_df["Gender"].count()
# #male_unique = male_df["ID"].unique()
# #male_unique_count = male_unique.count()
# male_avg_price = male_df["Price"].mean()
# # male_tot_purch_value =  male_df["Price"].sum()
# #male_avg_total = male_tot_purch_value / male_total
# print(male_total)
# #print(male_unique_count)
# print(male_avg_price)
# print(male_tot_purch_value)
# print(male_avg_total)
# print(male_avg_price)
# #print(male_tot_purch_value)
# #print(male_avg_total)
# female_df = purchase_data.loc[(
#      purchase_data["Gender"] == "Female")]
# female_total = female_df["Gender"].count()
# female_avg_price = female_df["Price"].mean()
# female_tot_purch_value =  female_df["Price"].sum()
# female_avg_total = female_tot_purch_value / female_total
# print(female_total)
# print(female_avg_price)
# print(female_tot_purch_value)
# print(female_avg_total)
#non_gender_specific = Total - male_total - female_total
# rows = df[(df[column1] <= dict[column1]) & (df[column2] <= dict[column2])]
# non_g_s_df = purchase_data.loc[(
# purchase_data[(["Gender"] != "Male") & (purchase_data["Gender"] != "Female"))]

# summary_table_purch_analysis = pd.DataFrame({"Average Price by Gender": male_avg_price,
#                                               "Purchase Count": male_total})
# summary_table_purch_analysis
#print(purch_count)

Total Purchase Value Avg Total Purchase per Person  \
Gender                                                                     
Female                             3.6e+02                          0.63   
Male                                 2e+03                           3.4   
Other / Non-Disclosed                5e+01                         0.087   

                      Average Purchase Price  
Gender                                        
Female                                  $3.2  
Male                                    $3.0  
Other / Non-Disclosed                   $3.3

## Age Demographics

* Establish bins for ages
* Categorize the existing players using the age bins. Hint: use pd.cut()
* Calculate the numbers and percentages by age group
* Create a summary data frame to hold the results
* Optional: round the percentage column to two decimal points
* Display Age Demographics Table
	Total Count	Percentage of Players
    
<10	17	2.95
10-14	22	3.82
15-19	107	18.58
20-24	258	44.79
25-29	77	13.37
30-34	52	9.03
35-39	31	5.38
40+	12	2.08


In [303]:
# Create bins in which to place values based ages
bins = [0, 9, 14, 19, 24, 29,
        34, 39, np.inf]
#Create labels for these bins
group_labels = ["<10", "10 to 14", "15-19", "20-24", "25-29", "30-34",
                "35-39", "40+"]
# Place the data series into a new column inside of the DataFrame
purchase_data['Age_Summary'] = pd.cut(purchase_data['Age'], bins, labels=group_labels)
# purchase_data.head()

#filter by SN to eliminate duplicate purchasers: 
clean_age_df = purchase_data.drop_duplicates(subset='SN', keep="last")
age_Total = clean_age_df["SN"].count()
#print('The total Number of Players is', age_Total)
#clean_age_df.head()

# Using GroupBy in order to separate the data into fields according to "gender" values
age_group = clean_age_df.groupby(['Age_Summary'])
# age_group = clean_age_df.groupby("Age_Summary")
#age_group_df.head()

age_group = clean_age_df["Age_Summary"].value_counts()
age_group.head(10)


#Percentage of Players
Percentage_Players = age_group/ age_Total
#type(Percentage_Players)
# print(Percentage_Players)
# age_sorted_df = age_group.sort_values("Age_Summary", ascending=False)
# #Percentage_Players.head(10)
# age_sorted_df.head(10)
##########################
#Creating a new DataFrame using both duration and count
age_summary_table = pd.DataFrame({"Total Count": age_group,
                                   "Percentage of Players": Percentage_Players
                                  })
age_summary_table["Percentage of Players"] = age_summary_table["Percentage of Players"].map( 
       "{:.2%}".format)
age_summary_table.head(10)

##########################

#Percentage of Players
# Percentage_Players = age_group["SN"].count() / filter_sn_unique
# type(Percentage_Players)
# print(Percentage_Players)

#Creating a new DataFrame using both duration and count
# age_summary_table = pd.DataFrame({"Age_Summary": age_group,
#                                 "Percentage_Players": Percentage_Players})
# age_summary_table["Percentage_Players"] = age_summary_table["Percentage_Players"].map( 
#      "{:.2%}".format)
# age_summary_table.head()

# age_merge_table = pd.merge(Percentage_Players, age_group, on="Age_Summary")
# age_merge_table
#pd.cut(purchase_data["Age"], bins, labels=group_labels).head()

#purchase_data_group = pd.DataFrame(purchase_data["Age_Summary"].value_counts())

#clean_ages = purchase_data.drop_duplicates(["SN", "Age"])[["SN", "Age"]]
#Total_ages = clean_ages["SN"].count()
#clean_ages.head()
#Total_ages
#purchase_data_group_sort = purchase_data_group.groupby("Age_Summary")
#purchase_data.head()
#purchase_data_group_sort.head()
#df["Age_Summary"] = pd.cut(df["Age"], bins, labels=group_labels).head()

# Creating a group based off of the bins
# df = clean_ages.groupby('SN')
# df['Age_Summary'].count().head(10)


Total Count Percentage of Players
20-24             258                44.79%
15-19             107                18.58%
25-29              77                13.37%
30-34              52                 9.03%
35-39              31                 5.38%
10 to 14           22                 3.82%
<10                17                 2.95%
40+                12                 2.08%

## Purchasing Analysis (Age)

* Bin the purchase_data data frame by age
* Run basic calculations to obtain purchase count, avg. purchase price, avg. purchase total per person etc. in the table below
* Create a summary data frame to hold the results
* Optional: give the displayed data cleaner formatting
* Display the summary data frame
	Purchase Count	Average Purchase Price	Total Purchase Value	Avg Total Purchase per Person
    
10-14	28	$2.96	$82.78	$3.76
15-19	136	$3.04	$412.89	$3.86
20-24	365	$3.05	$1,114.06	$4.32
25-29	101	$2.90	$293.00	$3.81
30-34	73	$2.93	$214.00	$4.12
35-39	41	$3.60	$147.67	$4.76
40+	13	$2.94	$38.24	$3.19
<10	23	$3.35	$77.13	$4.54

In [304]:
#purchase_data.head()
purchase_count_age = purchase_data['Age_Summary'].value_counts()
#purchase_count_age.head(10)
##############

# # Using GroupBy in order to separate the data into fields according to "state" values
grouped_SN_df = purchase_data.groupby(['SN'])

# # In order to be visualized, a data function must be used...
grouped_SN_df.count().head(10)

#Creating a new DataFrame using both duration and count
# pur_summary_table = pd.DataFrame({"	Purchase Count": purchase_count_age,
# # #                                     "Percentage of Players": Percentage_Players
#                                   })
# # pur_summary_table["Percentage of Players"] = pur_summary_table["Percentage of Players"].map( 
# #       "{:.2%}".format)
# pur_summary_table.head(10)

######################################################

#age_purchase_count = age_group['Age_Summary'].count
# #df.loc[df['mID']=='A','hID'].agg(['nunique','count','size'])
# #print(age_purchase_count)
# rev_purchanalysis = age_group.groupby( ["Age_Summary","Purchase ID"] ).count()
# #rev_purchanalysis = age_group.set_index('Age_Summary', inplace=True)
# rev_purchanalysis.head()
#print(rev_purchanalysis)
# new_group = purchase_data.groupby("Age_Summary")
# new_group.head()
# count_pd = pd.DataFrame(new_group, columns=[
#                         "Age_Summary", "Purchase Count"])
# Find how many rows fall into each bin
# print(new_group["Age_Summary"].count())

# new_group[["Price"]].mean()

# merge_table = pd.merge(new_group, count_pd, on="customer_id")
# merge_table

Purchase ID  Age  Gender  Item ID  Item Name  Price  Age_Summary
SN                                                                            
Adairialis76            1    1       1        1          1      1            1
Adastirin33             1    1       1        1          1      1            1
Aeda94                  1    1       1        1          1      1            1
Aela59                  1    1       1        1          1      1            1
Aelaria33               1    1       1        1          1      1            1
Aelastirin39            2    2       2        2          2      2            2
Aelidru27               1    1       1        1          1      1            1
Aelin32                 3    3       3        3          3      3            3
Aelly27                 2    2       2        2          2      2            2
Aellynun67              1    1       1        1          1      1            1

## Top Spenders

* Run basic calculations to obtain the results in the table below
* Create a summary data frame to hold the results
* Sort the total purchase value column in descending order
* Optional: give the displayed data cleaner formatting
* Display a preview of the summary data frame
	Purchase Count	Average Purchase Price	Total Purchase Value
SN			
Lisosia93	5	$3.79	$18.96
Idastidru52	4	$3.86	$15.45
Chamjask73	3	$4.61	$13.83
Iral74	4	$3.40	$13.62
Iskadarya95	3	$4.37	$13.10


## Most Popular Items

In [305]:
user_counts = purchase_data["SN"].value_counts()
#user_counts.head()

# Using GroupBy in order to separate the data into fields according to "state" values
grouped_pur_df = purchase_data.groupby(['SN'])

# In order to be visualized, a data function must be used...
#grouped_pur_df.count().head(10)

total_sn = grouped_pur_df["Price"].sum()
#total_sn.head()

###############################
Ave_PurPrice = user_counts / total_sn

#Creating a new DataFrame using both duration and count
spender_summary_table = pd.DataFrame({"Purchase Count": user_counts,
                                       "Total Purchase Value": total_sn,
                                       "Average Purchase Price": Ave_PurPrice,
# # #                                     "Average Purchase Price": male_avg_price
                                      })
spender_summary_table["Average Purchase Price"] = spender_summary_table["Average Purchase Price"].map("${:.2f}".format)
spender_summary_table["Total Purchase Value"] = spender_summary_table["Total Purchase Value"].map("${:.2f}".format)
#spender_summary_table.head() 
sort_spend = spender_summary_table.sort_values("Purchase Count", ascending=False)
sort_spend.head() 
# SN by MAx purchase count
# purchase_data.groupby(['SN']).groups.agg({
#  })

# topspenders_df = user_data.sort_values("Purchase ID", ascending=False)
# topspenders_df.head()
#user_data.head()
# #sn_avg = purchase_data.groupby('SN')['Price'].unique()
# #sn_sum = purchase_data.groupby('SN')['Price'].sum()
# max_sn_sum = sn_sum.max()
# print(sn_avg)
# #print(max_sn_sum)

# df = purchase_data[['SN','Price']].groupby(['SN'])['Price'] \
#                              .count() \
#                              .reset_index(name='count') \
#                              .sort_values(['count'], ascending=False) \

# #print (df)
#AVG price of these purchaes
#Total purchase value
#sort_spend.dtypes

Purchase Count Total Purchase Value Average Purchase Price
Lisosia93                 5               $18.96                  $0.26
Iral74                    4               $13.62                  $0.29
Idastidru52               4               $15.45                  $0.26
Asur53                    3                $7.44                  $0.40
Inguron55                 3               $11.11                  $0.27

* Retrieve the Item ID, Item Name, and Item Price columns


* Group by Item ID and Item Name. Perform calculations to obtain purchase count, item price, and total purchase value


* Create a summary data frame to hold the results


* Sort the purchase count column in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the summary data frame



In [318]:
#purchase_data.columns
# Extract columns 
reduced_item_pd = purchase_data.iloc[:, [
     4, 5, 6]]
#reduced_item_pd.head()
# # It is also possible to group a DataFrame by multiple columns
# # This returns an object with multiple indexes, however, which can be harder to deal with
grouped_item_data = reduced_item_pd.groupby(['Item ID', 'Item Name'])

grouped_item_data.count().head(5)
itemcount = grouped_item_data['Item ID'].count()
#print(itemcount)  

tot_price = grouped_item_data['Price'].sum()
#print(tot_price)

item_price = grouped_item_data['Price'].unique()

########################################
#Creating a new DataFrame using both duration and count
item_summary_table = pd.DataFrame({"Purchase Count": itemcount,
                                       "Total Purchase Value": tot_price,
                                       "Price": item_price,
                                      })
#item_summary_table["Average Purchase Price"] = item_summary_table["Average Purchase Price"].map("${:.2f}".format)
item_summary_table["Total Purchase Value"] = item_summary_table["Total Purchase Value"].map("${:.2f}".format)
#item_summary_table.head() 
sort_item = item_summary_table.sort_values("Purchase Count", ascending=False)
sort_item.head() 
# TotalPurchaseValue = itemcount * grouped_item_data.iloc["Price"]
# print(TotalPurchaseValue)
#sort_item.dtypes

Purchase Count  \
Item ID Item Name                                                      
178     Oathbreaker, Last Hope of the Breaking Storm              12   
145     Fiery Glass Crusader                                       9   
108     Extraction, Quickblade Of Trembling Hands                  9   
82      Nirvana                                                    9   
19      Pursuit, Cudgel of Necromancy                              8   

                                                     Total Purchase Value  \
Item ID Item Name                                                           
178     Oathbreaker, Last Hope of the Breaking Storm               $50.76   
145     Fiery Glass Crusader                                       $41.22   
108     Extraction, Quickblade Of Trembling Hands                  $31.77   
82      Nirvana                                                    $44.10   
19      Pursuit, Cudgel of Necromancy                               $8.16   

                                                       Price  
Item ID Item Name                                             
178     Oathbreaker, Last Hope of the Breaking Storm  [4.23]  
145     Fiery Glass Crusader                          [4.58]  
108     Extraction, Quickblade Of Trembling Hands     [3.53]  
82      Nirvana                                        [4.9]  
19      Pursuit, Cudgel of Necromancy                 [1.02]

## Most Profitable Items

* Sort the above table by total purchase value in descending order


* Optional: give the displayed data cleaner formatting


* Display a preview of the data frame



In [317]:
sort_item_pur = sort_item.sort_values("Total Purchase Value", ascending=False)
sort_item_pur.head() 

Purchase Count  \
Item ID Item Name                                           
63      Stormfury Mace                                  2   
29      Chaos, Ender of the End                         5   
173     Stormfury Longsword                             2   
1       Crucifer                                        3   
38      The Void, Vengeance of Dark Magic               4   

                                          Total Purchase Value   Price  
Item ID Item Name                                                       
63      Stormfury Mace                                   $9.98  [4.99]  
29      Chaos, Ender of the End                          $9.90  [1.98]  
173     Stormfury Longsword                              $9.86  [4.93]  
1       Crucifer                                         $9.78  [3.26]  
38      The Void, Vengeance of Dark Magic                $9.48  [2.37]